services/fee-estimator/shared/utils/dynamicFees.js

```diff
@@ -13,42 +13,110 @@
-	logger.debug(`Estimating fees with 'α' for EMA set to ${alpha}.`);
+	const cachedFeeEstimate = await cacheRedisFees.get(cacheKey);
 
-	const feeEst = {};
-	if (Object.keys(prevFeeEstPerByte).length === 0) {
-		prevFeeEstPerByte = { low: 0, med: 0, high: 0 };
-	}
+	const cachedFeeEstimateHeight = !cacheKey.includes('Quick') && cachedFeeEstimate
+		? cachedFeeEstimate.blockHeight : 0; // 0 implies does not exist
+
+	const prevFeeEstPerByte = fromHeight > cachedFeeEstimateHeight
+		? { blockHeight: fromHeight - 1, low: 0, med: 0, high: 0 } : cachedFeeEstimate;
+
+	const range = size => Array(size).fill().map((_, index) => index);
+	const feeEstPerByte = {};
+	const blockBatch = {};
+	do {
+		/* eslint-disable no-await-in-loop */
+		const idealEMABatchSize = config.feeEstimates.emaBatchSize;
+		const finalEMABatchSize = (() => {
+			const maxEMABasedOnHeight = prevFeeEstPerByte.blockHeight - genesisHeight;
+			if (idealEMABatchSize > maxEMABasedOnHeight) return maxEMABasedOnHeight + 1;
+			return idealEMABatchSize;
+		})();
+
+		blockBatch.data = await BluebirdPromise.map(
+			range(finalEMABatchSize),
+			async i => {
+				const { header, transactions } = await requestConnector(
+					'getBlockByHeight',
+					{ height: prevFeeEstPerByte.blockHeight + 1 - i },
+				);
+				return { ...header, transactions };
+			},
+			{ concurrency: 50 },
+		);
 
-	Object.keys(feePerByte).forEach((property) => {
-		feeEst[property] = alpha * feePerByte[property] + (1 - alpha) * prevFeeEstPerByte[property];
-	});
+		Object.assign(prevFeeEstPerByte,
+			await getEstimateFeeByteForBlock(blockBatch, prevFeeEstPerByte));
 
-	const EMAoutput = {
-		feeEstLow: feeEst.low,
-		feeEstMed: feeEst.med,
-		feeEstHigh: feeEst.high,
-	};
-	return EMAoutput;
+		// Store intermediate values, in case of a long running loop
+		if (prevFeeEstPerByte.blockHeight < toHeight) {
+			await cacheRedisFees.set(cacheKey, prevFeeEstPerByte);
+		}
+
+		/* eslint-enable no-await-in-loop */
+	} while (toHeight > prevFeeEstPerByte.blockHeight);
+
+	Object.assign(feeEstPerByte, prevFeeEstPerByte);
+	await cacheRedisFees.set(cacheKey, feeEstPerByte);
+
+	logger.info(`Recalulated dynamic fees: L: ${feeEstPerByte.low} M: ${feeEstPerByte.med} H: ${feeEstPerByte.high}`);
+
+	return feeEstPerByte;
 };
 
+const checkAndProcessExecution = async (fromHeight, toHeight, cacheKey) => {
+	let result = await cacheRedisFees.get(cacheKey);
+	if (!executionStatus[cacheKey]) {
+		try {
+			// If the process (full / quick) is already running,
+			// do not allow it to run again until the prior execution finishes
+			executionStatus[cacheKey] = true;
+			result = await getEstimateFeeByteForBatch(fromHeight, toHeight, cacheKey);
+		} catch (err) {
+			logger.error(err.stack || err.message);
+		} finally {
+			executionStatus[cacheKey] = false;
+		}
+	}
+	return result;
+};
+
+const isFeeCalculationRunning = (cacheKey) => executionStatus[cacheKey];
```

Comment:
Please rename and move it to the top as suggested. Also, kindly do not forget to refactor the code at other places accordingly.
```suggestion
};
```

Reply:
Applied.